hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/KeyProviderCache.java

```diff
@@ -68,8 +68,11 @@ public void onRemoval(
         })
         .build();
 
-    ShutdownHookManager.get().addShutdownHook(new KeyProviderCacheFinalizer(),
-        SHUTDOWN_HOOK_PRIORITY);
+    try {
+      ShutdownHookManager.get().addShutdownHook(new KeyProviderCacheFinalizer(), SHUTDOWN_HOOK_PRIORITY);
+    } catch (IllegalStateException e) {
+      LOG.warn("shutdownHook not added", e);
+    }
```

Comment:
We should first check if a shutdown is in progress and then add a shutdown hook if not, rather than catching the exception. Like the example in `FileSystem`:
https://github.com/apache/hadoop/blob/60e0fe87094771b3188b9b1cd4bf7acf54d40ab7/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L3715-L3720

Reply:
Thanks, fixed.